src/echo/echo.rs

```diff
@@ -36,119 +36,90 @@ static HELP: &str = r#"
+            Some(n) => ret = (ret << bits_per_digit) | n,
+            None => break,
+        }
+        input.next();
+    }
+    std::char::from_u32(ret)
 }
 
-fn convert_str(string: &[u8], index: usize, base: Base) -> (char, usize) {
-    let (max_digits, is_legal_digit): (usize, fn(u8) -> bool) = match base {
-        Base::B8 => (3, |c| (c as char).is_digit(8)),
-        Base::B16 => (2, |c| (c as char).is_digit(16)),
-    };
-
-    let mut bytes = vec![];
-    for offset in 0..max_digits {
-        if string.len() <= index + offset as usize {
-            break;
-        }
-        let c = string[index + offset as usize];
-        if is_legal_digit(c) {
-            bytes.push(c as u8);
-        } else {
-            break;
+pub fn unescape(input: &str) -> String {
+    let mut output = String::with_capacity(input.len());
+    let mut iter = input.chars().peekable();
+    while let Some(c) = iter.next() {
+        match c {
+            '\\' => match iter.next() {
+                Some(c) => match c {
+                    '\\' => output.push('\\'),
+                    'a' => output.push('\x07'),
+                    'b' => output.push('\x08'),
+                    'c' => break,
+                    'e' => output.push('\x1b'),
+                    'f' => output.push('\x0c'),
+                    'n' => output.push('\n'),
+                    'r' => output.push('\r'),
+                    't' => output.push('\t'),
+                    'v' => output.push('\x0b'),
+                    'x' => match parse_code(&mut iter, 16, 2, 4) {
+                        Some(c) => output.push(c),
+                        None => output.push_str("\\x"),
+                    },
+                    '0' => match parse_code(&mut iter, 8, 3, 3) {
+                        Some(c) => output.push(c),
+                        None => output.push_str("\\0"),
+                    },
+                    _ => {
+                        output.push('\\');
+                        output.push(c);
+                    },
+                },
+                None => output.push(c),
+            },
+            _ => output.push(c),
         }
     }
+    output
+}
 
-    if bytes.is_empty() {
-        (' ', 0)
-    } else {
-        (
-            usize::from_str_radix(from_utf8(bytes.as_ref()).unwrap(), base as u32).unwrap() as u8
-                as char,
-            bytes.len(),
-        )
+#[cfg(test)]
+mod tests {
+    #[test]
+    fn unescape() {
+        assert_eq!("\n", super::unescape("\\n"));
+        assert_eq!("foo", super::unescape("foo\\cbar"));
+        assert_eq!("foo\n bar", super::unescape("foo\\xa bar"));
+        assert_eq!("foo bar", super::unescape("foo\\x20bar"));
+        assert_eq!("foo bar", super::unescape("foo\\040bar"));
+        assert_eq!("foo bar", super::unescape("foo\\0040bar"));
+        assert_eq!("foo\\x bar", super::unescape("foo\\x bar"));
+        assert_eq!("foo\\0 bar", super::unescape("foo\\0 bar"));
```

Comment:
If these weren't being tested before, do you mind adding these to the `echo` tests in the `tests` directory?

Reply:
Oh so there are tests for this command already. :P

Looks like none of the irregular escape sequences are tested there. Will add.